### PR TITLE
EES-4187 Use UniversalWrapper with caching code

### PR DESCRIPTION
**Pull request by markhiveit**
This PR adds the ability to run async code without blocking threads when using caching - specifically caching code in and around `CacheAspect`/`CacheAttribute`. We were having thread exhaustion / deadlock issues when there was a large number of simultaneous cache misses. See Jira comments for EES-4177 and other related tickets for a more detailed investigation, recreation, and breakdown of the issue.

UniversalWrapper allows us to define `WrapSync` and `WrapAsync` methods for cache attributes. I believe the code in these methods is injected into the attributed method - with the `WrapSync` code being injected if the attributed method is synchronous, and the `WrapAsync` code being injected if the attributed method is async. To take advantage of this, we've moved most code that was previously in `CacheAspect` to `CacheAttribute`. It has also meant that `MemoryCacheAttribute` and `BlobCacheAttribute` now have separate `Get/GetAsync` and `Set/SetAsync` methods.

`MemoryCacheService`'s calls to get and set cached values are not async, therefore we can still support adding the `MemoryCache` attribute to sync and async methods. 

BlobCache, however, requires async calls to blob storage, so we cannot allow the BlobCache attribute to be applied to sync methods without blocking threads.

With async code now being handled in `WrapAsync`, there is no longer any need to manage `Task`s in `ReflectionExtensions#TryBoxToResult` and `TryUnboxResult`.

A few other changes have been made:

### Removal of `CacheAttribute`'s `Priority` and `AllowMultiple`

`BaseUniversalWrapperAspect#BaseHandle` doesn't provide support for allowing multiple cache triggers to be supplied with differing priorities. I believe to make this work we'd need to pull `BaseHandle` into our codebase. Fixing this seems nontrivial, and we haven't actually used this priority functionality, so I removed it.

### Failing unit test related to MemoryCache's `OverrideDurationInSeconds` and `OverrideExpirySchedule`

A unit test - `MemoryCacheAttributeTests#OverrideDurationInSecondsAndExpirySchedule` - to test the ability to override a memory cache's configuration was failing when run alongside all other unit tests, but passing when run by itself. We determined that this is because it was reusing an instance of `MemoryCacheAttribute` and therefore the constructor wasn't being called - the constructor previously set `DurationInSections` and `ExpirySchedule` to the value of the static override variables, if they weren't null. So now instead we check whether we should use the static override values whenever we call `service.SetItem`.

I also renamed the private static override variables because Rider told me to.

### Local testing

I created a performance test `getReleasePageCacheMiss.test.ts`. This is a copy of `getReleasePage.test.ts`, but of shorter duration with more virtual users, which caused thread exhaustion before this PR's changes. I used this locally alongside `dotnet-counters` to prove the issue had been solved.

### What this PR doesn't do

If a lot of cache misses happen all at once, the code will still generate and write the cache file multiple times.

### Credit

Most of this was worked out by Ben - you can see his test project projects to recreate the issue and then prove that universal wrapper solves the issue here:
- https://github.com/benoutram/cache-aspect-test
- https://github.com/benoutram/cache-aspect-universal-wrapper

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Common/Cache/BlobCacheAttribute.cs
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Common/Cache/BlobCacheAttribute.cs:54]: Reading this comment sounds like there's an actual limitation here, which there currently is, but it's because these methods need to be written with synchronous `GetItem` and `SetItem` calls to the `IBlobCacheService` which we've not implemented.

But we could!

I've shown in the sample project that this is do-able without causing blocking if we did this. We'd need to do it throughout the chain of execution by also adding synchronous versions of a few methods to `IBlobStorageService` to allow this.

`BlobCacheAttribute` and `IBlobCacheService` are intended to be generic and not tied to any particular type of blob storage.

Where the cache resides depends on the type of `IBlobStorageService` wired in to them. In our case it's an Azure blob storage service.

Whether or not a specific type of `IBlobStorageService` can implement synchronous methods depends on if the client library offers that support (the Azure one does), and if we decide to go ahead and implement them (we haven't needed to).

We could change `IBlobCacheService` to inherit from `ICacheService` which already defines `GetItem` and `SetItem`, and make a similar change to `IBlobStorageService` to have synchronous methods required by `BlobCacheService` and update `BlobCacheService` to use them.

We'd then have a richer `BlobCacheAttribute` implementation that doesn't have this async-only limitation.

The limitation in annotating synchronous methods then becomes the fact that we've not yet fully implemented our Azure storage service yet.

It would work as soon as we do, and will work with an alternate storage service if one is implemented.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAttribute.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAttribute.cs:29]: I discussed this with Dunc. I won't speak for Dunc, but after that conversation my thought was as it wasn't being used, we can remove it and create a ticket to add it again if we feel strongly about it.

I have also wondered if we can basically replicate the same effect - not sure if this would actually do the same job though:
```
[MemoryCache]
public T method1()
{
  method2();
}

[BlobCache]
public T method2()
{
  DoStuff();
}
```

File: src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs:61]: You mean remove the exception? i.e. no case statement for Tasks? I think maybe I'm just being overly cautious by throwing this exception. I think I'd keep it for now - but happy to go along with what others think.

I think it's right that the box/unbox things shouldn't deal with Tasks. My intuition now is that it's dangerous to let them do so - I think you'd agree with that but just saying to make sure we're on the same page.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs:132]: My intuition is it's potentially dangerous to allow it to deal with Tasks - and we avoid possible problems by throwing exceptions on them. It shouldn't be difficult for a method to await a task before passing it to these methods - and if they can't await then I'm not sure that method should be dealing with a task.

Maybe there is another possible use case where handling tasks is helpful or something I'm missing?

File: src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs:98]: Would provide a better message than this e.g.

```cs
throw new ArgumentException(
    "Cannot unbox Tasks as this may cause deadlock. Consider awaiting the result first.");
```

File: src/GovUk.Education.ExploreEducationStatistics.Common/Cache/BlobCacheAttribute.cs
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Cache/BlobCacheAttribute.cs:54]: Agree with Ben here, it might be a good idea to just fully implement both sync and async to avoid any surprises.

File: src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/CacheAspectTests.cs
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/CacheAspectTests.cs:1284]: It'd be good if we could implement some unit or integration tests that verify that the previous deadlocking behaviour is no longer occurring.

If we make any more changes to the caching code in the future, it'd be good to know that we aren't introducing any regressions. Getting feedback at this stage would be much better than relying on the performance tests (esp. as we don't run them all the time).

I think @benoutram was leaning more towards integration tests as they probably give a more accurate reflection of how this would work in production.

Not entirely sure what we would need to do here, but would suggest we split out a separate task for this as it could get pretty gnarly.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAttribute.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAttribute.cs:94]: This cast previously wasn't necessary because we were returning inside the handle method, which has the return type `object`. Now that this code is inside a WrapAsync/WrapSync methods, which return `T`, we now need the cast.

I think to fix this we need the out param type of `TryBoxToResult` to be dynamic, depending on what type is passed as an argument to it, and I don't believe we can do that.

If there is a way to avoid this cast, I'm not sure what it is. `TryBoxToResult` checks that `cachedResult` is an instance of type `T`, but I guess we could also do this check in WrapSync/WrapAsync to protect against future refactors of `TryBoxToResult`, but I'm inclined to think it's `TryBoxToResult`'s unit tests' responsibility to protect against this. So I'm personally inclined to leave this.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAttribute.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAttribute.cs:29]: I started implementing this, and I think I've spotted a bug with the existing code.

I believe the bug is around how `ForceUpdate` interacts with multiple cacheTriggers. If you have multiple triggers, the first priority being `MemoryCache` with `ForceUpdate` set, and the second being `BlobCache` with a cached value, currently, the `MemoryCache` wouldn't get updated.

I think there is some ambiguity around what we would expect `ForceUpdate` to do in different situations, so if `MemoryCache` is prioritised and has a cached value, but the lower priority `BlobCache` does have `ForceUpdate` set, we will probably want to update the `BlobCache` cached value, but should we just return the `MemoryCache` cached value, or should that be updated as well?

Rather than try to address this in this PR, I'm going to create another ticket for it.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAttribute.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAttribute.cs:130]: See other comment

File: src/GovUk.Education.ExploreEducationStatistics.Common/Cache/BlobCacheAttribute.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Cache/BlobCacheAttribute.cs]: I've made some changes 
- `IBlobCacheService` now inherits `ICacheService<IBlobCacheKey>` 
- `BlobCacheAttribute` now has Get/GetAsync/Set/SetAsync methods

I've not gone all the way and implemented `GetItem` and `SetItem` in `BlobCacheService` though - instead they throw `NotImplementedException`s. Since we're not using the `BlobCache` attribute on a sync method currently, and it's not clear we will in the future, I don't think spending the time necessary to implement and test this is currently justified.

IIRC, when previously discussing this with Ben on Slack, this was the approach he initially suggested. But I'm happy to complete the implementation if I've got that wrong and/or people are keen on me actually implementing the functionality.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs:61]: I'd still prefer to throw an exception because I think it prevents a lot of potential mischief - see comment below - but happy to change it if others want.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs:98]: Also changed the exception message in `TryBoxToResult`.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs]: Again, would prefer that we change this. The code isn't unsafe as we're not calling any synchronous methods, so the method should do what it says on the tin i.e. box it up.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs:61]: Repeat of another comment - would reinstate the previous behaviour as this actually unsafe as we're not calling any synchronously blocking methods like `Result`. All test relevant test cases should also be reinstated.

Also this exception doesn't even make sense as we're not unboxing the result.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs:132]: Happy to keep the exception for now. I think in the future I'd probably still like to see this converted into an asynchronous method so we can unbox it though.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAttribute.cs
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAttribute.cs:29]: Yeh I was alluding that there are potential issues with the existing code, so it's fine to leave this to another ticket!

File: src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs]: Restored original code and restored associated `TryBoxToResult` unit tests.

File: src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/CacheAspectTests.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/CacheAspectTests.cs:1352]: I've renamed it as suggested. This ended up breaking a couple of unit tests - `NoParams_TaskActionResult` and `NoParams_NestedTaskEither`. I think the fixes I've made are okay, but you will probably want to look at them specifically. 

It is a little confusing as we've got a sync method that returns a `Task<...>` and it seems universal wrapper calls `WrapAsync` if such a method is annotated. I'm not sure this situation of non-async method returning a `Task<...>` should ever happen in actual code, but I've left the tests in.

File: src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/CacheAspectTests.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/CacheAspectTests.cs:719]: Made this change, and there were also a couple of other instances where I refactored to use the various `AssertCache*` methods.

File: src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/MemoryCacheAttributeTests.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/MemoryCacheAttributeTests.cs:364]: I've removed a few `Assert.IsType`s that were immediately followed by `Assert.Equal`s.

**Review by ntsim — APPROVED** [review]
Approved, but would let Ben give it a look before merging as mentioned in standup.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Cache/BlobCacheAttribute.cs
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Cache/BlobCacheAttribute.cs]: This was as far as I'd envisaged the non-async path getting implemented so it's all good with me.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs:132]: I think it's a good idea, I've left another comment on the lines in the code just above this where the exception gets raised to remind us to create ticket or at least add a TODO to the code so we don't forget this.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs]: Created EES-4268 and added the TODO. You might want to update the ticket - I wasn't sure what to put in the description.

**Comment by markhiveit** [comment]
I believe this PR is good to be merged now?

**Pull request merged**
